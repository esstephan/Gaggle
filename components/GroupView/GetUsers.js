import React, { Component } from 'react';
import {
  AppRegistry,
  StyleSheet,
  Text,
  TextInput,
  View
} from 'react-native';
import GroupView from './GroupView.js';

export default GetUsers = function() {
  var url = 'https://catfacts-api.appspot.com/api/facts?number=5';

  return fetch(url)
    .then((response) => response.json())
    .then((responseJson) => {
      this.setState({
        users: responseJson.facts //update when we move to firebase data
      })
    })
    .catch((error) => {
      console.error(error);
    });
};