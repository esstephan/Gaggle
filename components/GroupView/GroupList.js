import React from 'react';
import { TouchableOpacity, View } from 'react-native';
import { Text, Icon } from 'native-base';

const GroupList = ({ _handleChangePage, userGroups, deleteGroup, uid }) => {
  return (
    <View
      style={{
        flex: 1,
        flexDirection: 'row',
        justifyContent: 'space-between',
        paddingTop: 30,
        flexWrap: 'wrap',
      }}
    >
      {userGroups.map((group, i) => (
        <TouchableOpacity
          key={i}
          style={{
            shadowColor: 'black',
            shadowOpacity: 0.8,
            shadowRadius: 3,
            shadowOffset: {
              height: 0,
              width: 0,
            },
            backgroundColor: 'white',
            margin: 20,
            width: 100,
            height: 100,
          }}
          onPress={() => _handleChangePage(group || '')}
        >
          <Text
            style={{
              
            }}
          >{group}</Text>
          <Icon
            name={'trash'}
            style={{
              color: 'red',
              textAlign: 'center',
            }}
            onPress={() => deleteGroup(uid, group)}
          />
        </TouchableOpacity>
        ))
      }
    </View>
  );
};

GroupList.propTypes = {
  _handleChangePage: React.PropTypes.func.isRequired,
  userGroups: React.PropTypes.array.isRequired,
  deleteGroup: React.PropTypes.func.isRequired,
  uid: React.PropTypes.string.isRequired,
};

export default GroupList;
