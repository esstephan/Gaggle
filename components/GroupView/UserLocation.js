/* eslint no-console: 0 */
'use strict';


var React = require('react');
var ReactNative = require('react-native');
var {
  StyleSheet,
  Text,
  View,
} = ReactNative;


export default class UserLocation extends React.Component {
  state = {
    //initialPosition: 'unknown',
    lastPosition: 'unknown',
  };

  watchID: ?number = null;

  componentDidMount() {
    this.watchID = navigator.geolocation.watchPosition((position) => {
      //var lastPosition = JSON.stringify(position);
      var lastPosition = [position.coords.longitude, position.coords.latitude];
      this.setState({lastPosition});
      //console.log(JSON.parse(this.state.lastPosition));
      //console.log(position);
    });
  }

  componentWillUnmount() {
    navigator.geolocation.clearWatch(this.watchID);
  }

  render() {
    return (
      <View>
        <Text>
          <Text style={styles.title}>Initial position: </Text>
          {this.state.initialPosition}
        </Text>
        <Text>
          <Text style={styles.title}>Current position: </Text>
          {this.state.lastPosition}
        </Text>
        <Text>
          <Text style={styles.title}>Longitude: </Text>
          {this.state.lastPosition[0]}
        </Text>
        <Text>
          <Text style={styles.title}>Latitude: </Text>
          {this.state.lastPosition[1]}
        </Text>
      </View>
    );
  }
}

var styles = StyleSheet.create({
  title: {
    fontWeight: '500',
  },
});