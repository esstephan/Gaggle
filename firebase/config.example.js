export default {
  apiKey: '',
  authDomain: '',
  databaseURL: '',
  storageBucket: '',
  messagingSenderId: '',
};
