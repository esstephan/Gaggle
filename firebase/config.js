import firebase from 'firebase';

let config = {
  apiKey: "",
  authDomain: "",
  databaseURL: "",
  storageBucket: "",
  messagingSenderId: ""
};

export default firebaseRef = firebase.initializeApp(config);