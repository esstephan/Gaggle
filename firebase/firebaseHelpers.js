import firebase from 'firebase';
import config from './config';

export const firebaseRef = firebase.initializeApp(config);
export const firebaseDB = firebaseRef.database();


export const addUserToGroup = (userObj, groupName) => {
  const user = {
    displayName: userObj.displayName,
    location: userObj.location,
  };

  firebaseDB.ref(`groups/${groupName}/members/${userObj.uid}`).set(user);
  firebaseDB.ref(`users/${userObj.uid}/groups`).push(groupName);
};

export const removeUserFromGroup = (uid, groupName) => {
  console.log(uid, `group name`, groupName);
  console.log('firebase ref', firebaseDB.ref(`users/${uid}/groups`));

  firebaseDB.ref(`users/${uid}/groups`);
  // firebaseDB.ref(`groups/${groupName}/members/${uid}`).remove();
}

export const getAllGroupsInUser = (uid) => {
  const arrayOfGroups = [];

  firebaseDB.ref(`users/${uid}/groups`).on('value', (snapshot) => {
    const groupsObj = snapshot.val();

    for (let key in groupsObj) {
      arrayOfGroups.push(groupsObj[key]);
    }
  }, (error) => { console.log(`Error getting groups ${error}`); });

  return arrayOfGroups;
};

export const getAllUsersInGroup = (groupName) => {
  firebaseDB.ref(`groups/${groupName}/members`).once('value')
    .then((snapshot) => {
      return snapshot.val();
    });
};

export const getCurrentUser = () => {
  return firebaseRef.auth().currentUser;
};

export const getCurrentUserId = () => {
  const user = firebaseRef.auth().currentUser;
  return user.uid;
};

export const updateUserLocation = (activeGroup) =>  {
  let location = {};
  navigator.geolocation.getCurrentPosition((position) => {
    location = position;
    const userID = getCurrentUserId();
    const user = getCurrentUser();

    let updates = {};
    updates[`users/${user.uid}/location/`] = location;
    updates[`groups/${activeGroup}/members/${user.uid}`] = {
      uid: user.uid,
      displayName: user.displayName,
      location: location,
    };

    firebaseDB.ref().update(updates)
      .then(() => {
        console.log('Location update successful!');
        return location;
      })
      .catch((error) => { console.log(`Location update Error ${error}`); });
  },
    // Relates to geolocation position initiated above
    (error) => alert(JSON.stringify(error)),
    { enableHighAccuracy: true, timeout: 20000, maximumAge: 1000 },
  )
};

// Moved directly into MapDisplay to setState on the component whenever value is updated
// export const getMemberLocations = (activeGroup) => {
//   const locArray = [];
//   firebaseDB.ref(`groups/${activeGroup}/members/`).on('value', (snapshot) => {
//     snapshot.forEach((childSnapshot) => {
//       locArray.push(childSnapshot.val());
//     });
//   });
//  console.log('getMemberLocations LOCATIONS ARRAY', locArray);
//  return locArray;
// };
export const getGroupMemberLocations = groupName => (
  new Promise((resolve, reject) => {
    const locationsArray = [];
    firebaseDB.ref(`groups/${groupName}/members/`).once('value')
    .then((snapshot) => {
      snapshot.forEach((childSnapshot) => {
        const memberLocationObj = childSnapshot.val().location.coords;
        const location = [memberLocationObj.latitude, memberLocationObj.longitude];
        locationsArray.push(location);
      });
    })
    .then(() => {
      resolve(locationsArray);
    })
    .catch(err => reject(err));
  })
);
