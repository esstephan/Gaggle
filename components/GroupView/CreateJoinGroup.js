import React, { Component } from 'react';
import { Input, Button, Text, Item, View } from 'native-base';
import { addUserToGroup } from '../../firebase/firebaseHelpers';

export default class CreateJoinGroup extends Component {
  constructor(props) {
    super(props);
    this.state = {
      user: this.props.user,
    };
  }

  render() {
    return (
      <Item>
        <Input
          ref={(component) => { this._groupInput = component; }}
          placeholder="Group Name"
          autoCapitalize={'none'}
          style={{
            paddingLeft: 15,
          }}
        />
        <Button
          small
          style={{
            margin: 10,
          }}
          onPress={
            () => addUserToGroup(this.state.user, this._groupInput._root._lastNativeText)
          }
        >
          <Text>Create/Join</Text>
        </Button>
      </Item>
    );
  }

}

CreateJoinGroup.propTypes = {
  user: React.PropTypes.object.isRequired,
};
