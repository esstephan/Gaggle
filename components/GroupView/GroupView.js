import React, { Component } from 'react';
import { Alert, Text } from 'react-native';
import { Container, Header, Content } from 'native-base';
import { firebaseDB, updateUserLocation, getAllGroupsInUser, removeUserFromGroup } from '../../firebase/firebaseHelpers';
import GroupMapChat from '../GroupMapChat/GroupMapChat';
import CreateJoinGroup from './CreateJoinGroup';
import GroupList from './GroupList';
import Quiz from '../Quiz/Quiz';

export default class GroupView extends Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      user: this.props.user,
      users: [],
      activeGroup: 'Default',
    };

    this.groupsArray = getAllGroupsInUser(this.state.user.uid);
    this.usersRef = firebaseDB.ref('/users');
    this._handleChangePage = this._handleChangePage.bind(this);
    this.deleteGroup = this.deleteGroup.bind(this);
  }

  componentWillMount() {
    this._usersListener();
  }

  componentWillUnmount() {
    this.usersRef.off('value');
  }

  _handleChangePage(name) {
    this.setState({
      activeGroup: name,
    }, () => {
      updateUserLocation(this.state.activeGroup);
    });
    this.props.navigator.push({
      component: GroupMapChat,
      title: `${name} Group`,
      passProps: {
        user: this.props.user,
        groupName: name,
      },
      rightButtonTitle: 'Quiz',
      onRightButtonPress: () => this.props.navigator.push({
        component: Quiz,
        title: 'While You Wait...',
      }),
    });
  }

  deleteGroup(uid, groupName) {
    Alert.alert(
      'Delete Group',
      `Are you sure you want to remove yourself from ${groupName}?`,
      [
        { text: 'Cancel', onPress: () => console.log('Cancel pressed') },
        { text: 'Yes', onPress: () => removeUserFromGroup(uid, groupName), style: 'destructive' },
      ],
    );
  }

  _usersListener() {
    this.usersRef.on('value', (snapshot) => {
      const usersArray = [];
      snapshot.forEach((childSnapshot) => {
        usersArray.push(childSnapshot.val());
      });
      this.setState({ users: usersArray });
    }).bind(this);
  }

  render() {
    return (
      <Container>
        <Header />
        <Content>
          {this.groupsArray.length === 0 &&
            <Text
              style={{
                color: 'grey',
                textAlign: 'center',
                marginVertical: 10,
              }}
            >To join a group, just type in the name, and click Join!</Text>}
          <CreateJoinGroup user={this.state.user} />
          <GroupList
            _handleChangePage={this._handleChangePage}
            userGroups={getAllGroupsInUser(this.state.user.uid)}
            deleteGroup={this.deleteGroup}
            uid={this.state.user.uid}
          />
        </Content>
      </Container>
    );
  }
}

GroupView.propTypes = {
  navigator: React.PropTypes.object.isRequired,
  user: React.PropTypes.object.isRequired,
};