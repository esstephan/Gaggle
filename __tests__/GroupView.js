import 'react-native';
import React from 'react';
import GroupView from '../components/GroupView/GroupView.js';
import {firebaseRef, firebaseDB} from '../firebase/firebaseHelpers';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

// Test is failing:
//
// it('it renders correctly', () => {
//   const tree = renderer.create(
//     <GroupView />
//   );
//   expect(tree).toMatchSnapshot();
// });

//SEEMS TO DO THE SAME THING...

// test('test renders correctly', () => {
//   const tree = renderer.create(
//     <GroupView />
//   ).toJSON();
//   expect(tree).toMatchSnapshot();
// });

