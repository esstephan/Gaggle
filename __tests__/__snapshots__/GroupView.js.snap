exports[`test it renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "height": 1334,
    }
  }
  virtual={undefined}>
  <View>
    <View
      style={
        Object {
          "backgroundColor": "#F8F8F8",
          "borderBottomColor": "#a7a6ab",
          "borderBottomWidth": 0.5,
          "elevation": 3,
          "flexDirection": "row",
          "height": 64,
          "justifyContent": "center",
          "left": 0,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 15,
          "right": 0,
          "top": 0,
        }
      }
      virtual={undefined}>
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "alignSelf": "center",
            "flex": 1,
          }
        }
        virtual={undefined} />
      <View
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "center",
            "flex": 1,
          }
        }
        virtual={undefined} />
      <View
        style={
          Object {
            "alignItems": "flex-end",
            "alignSelf": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "flex-end",
          }
        }
        virtual={undefined} />
    </View>
  </View>
  <RCTScrollView>
    <View>
      <View />
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "backgroundColor": "#F8F8F8",
        "borderColor": "#cbcbcb",
        "borderTopWidth": 0.5,
        "elevation": 3,
        "flexDirection": "row",
        "height": 55,
        "justifyContent": "center",
        "left": 0,
        "right": 0,
      }
    }
    virtual={undefined}>
    <View
      style={
        Object {
          "alignSelf": "stretch",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
      virtual={undefined}>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hitSlop={undefined}
        isTVSelectable={true}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "center",
            "backgroundColor": "transparent",
            "borderBottomWidth": null,
            "borderColor": null,
            "borderLeftWidth": null,
            "borderRadius": 5,
            "borderRightWidth": null,
            "borderTopWidth": null,
            "elevation": 0,
            "flex": 1,
            "flexDirection": null,
            "height": 55,
            "justifyContent": "center",
            "opacity": 1,
            "paddingBottom": 6,
            "paddingLeft": 16,
            "paddingRight": 16,
            "paddingTop": 6,
            "shadowColor": null,
            "shadowOffset": null,
            "shadowOpacity": null,
            "shadowRadius": null,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail">
          Next Page
        </Text>
      </View>
    </View>
  </View>
</View>
`;
